<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>qDNAmod by zhixingfeng</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header class="without-description">
        <h1>qDNAmod</h1>
        <p></p>
        <p class="view"><a href="https://github.com/zhixingfeng/qDNAmod">View the Project on GitHub <small>zhixingfeng/qDNAmod</small></a></p>
        <ul>
          <li><a href="https://github.com/zhixingfeng/qDNAmod/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/zhixingfeng/qDNAmod/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/zhixingfeng/qDNAmod">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h1>
<a id="qdnamod" class="anchor" href="#qdnamod" aria-hidden="true"><span class="octicon octicon-link"></span></a>qDNAmod</h1>

<p>qDNAmod is a toolkit for quantitative detection of DNA modification heterogeneity from SMRT sequencing data. </p>

<h1>
<a id="system" class="anchor" href="#system" aria-hidden="true"><span class="octicon octicon-link"></span></a>System</h1>

<p>Linux</p>

<h1>
<a id="dependencies" class="anchor" href="#dependencies" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dependencies</h1>

<ol>
<li><p>R from <a href="http://www.r-project.org/">http://www.r-project.org/</a>
(Add directory containing binary file 'Rscript' to your enviroment variable PATH)</p></li>
<li><p>HDF5 from <a href="http://www.hdfgroup.org/HDF5/release/obtainsrc.html">http://www.hdfgroup.org/HDF5/release/obtainsrc.html</a>. 
(Specify enviroment varible HDF5_HOME as the directory of installed HDF5)</p></li>
</ol>

<h1>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h1>

<p>Unzip the downloaded file, enter the directory and type ./install.sh </p>

<p>Add (the unzipped directory of qDNAmod)/bin to your PATH</p>

<p>Set evironment variable QDNAMODHOME=(the unzipped directory of qDNAmod)</p>

<p>(g++ is required for installation)</p>

<p>Make sure your computer can access internet during the installation.</p>

<h1>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span class="octicon octicon-link"></span></a>License</h1>

<p>GPL2</p>

<h1>
<a id="input" class="anchor" href="#input" aria-hidden="true"><span class="octicon octicon-link"></span></a>Input</h1>

<p>The Input of qDNAmod is .cmp.h5 files (aligned reads).</p>

<h1>
<a id="example" class="anchor" href="#example" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example</h1>

<p><a href="http://bioinfo.au.tsinghua.edu.cn/member/zfeng/example_release.tar.gz">http://bioinfo.au.tsinghua.edu.cn/member/zfeng/example_release.tar.gz</a></p>

<h1>
<a id="maintainer" class="anchor" href="#maintainer" aria-hidden="true"><span class="octicon octicon-link"></span></a>Maintainer</h1>

<p>Zhixing Feng</p>

<h1>
<a id="contact" class="anchor" href="#contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Contact</h1>

<p><a href="mailto:zxfeng.thu@gmail.com">zxfeng.thu@gmail.com</a></p>

<h1>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h1>

<p>There are 3 steps: pileup reads, learn prior and detect modification proportion.</p>

<p>Step 1: pileup reads. Firstly, we need to pileup reads position by position.</p>

<pre><code>qDNAmod_pileup  [-r &lt;reagent&gt;] [-m &lt;mapQVthreshold&gt;] &lt;cmpH5file&gt; &lt;outdir&gt;

where:
    -r &lt;reagent&gt;,  --reagent &lt;reagent&gt;
        chemistry used for SMRT sequencing, candidates are: "C2", default is "C2"

    -m &lt;mapQVthreshold&gt;,  --mapQVthreshold &lt;mapQVthreshold&gt;
        minimal mapQV, default is 255

    &lt;cmpH5file&gt;
        (required)  aligned data in cmpH5 format

    &lt;outdir&gt;
        (required)  output directory
</code></pre>

<p>Step 2: learn prior. To reduce uncertainty of estimation, we need to roughly estimate IPD distribution of modified bases.</p>

<pre><code>qDNAmod_prior  &lt;native pileup dir&gt; &lt;WGA pileup dir&gt; &lt;outdir&gt;

where:

    &lt;native pileup dir&gt;
        (required)  native pileup data directory

    &lt;WGA pileup dir&gt;
        (required)  WGA pileup data directory

    &lt;outdir&gt;
        (required)  output directory
</code></pre>

<p>Step 3: detect modification proportion. Now we fit a Bayesian mixture model using prior estimated in step 2.</p>

<pre><code>qDNAmod_detect &lt;native pileup dir&gt; &lt;WGA pileup dir&gt; &lt;prior dir&gt; &lt;outdir&gt;

where:
    &lt;native pileup dir&gt;
        (required)  native pileup data directory

    &lt;WGA pileup dir&gt;
        (required)  WGA pileup data directory

    &lt;prior dir&gt;
        (required)  directory containing estimated prior

    &lt;outdir&gt;
        (required)  output directory
</code></pre>

<h1>
<a id="output" class="anchor" href="#output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output</h1>

<p>In the "outdir" of qDNAmod_detect, there are .txt files with the names "detect_(chromosome_name).txt" corresponding to results for (chromosome_name). 
In each .txt file, the columns are:</p>

<p>column 1 (locus): genome locus.</p>

<p>column 2 (strand): strand, 0 means forward, 1 means backward.</p>

<p>column 3 (prop): estimated modification proportion.</p>

<p>column 4 (N_1): expectation of number of kinetic variant bases.  </p>

<p>column 5 (N_0): expectation of number of normal bases. </p>

<p>column 6 (avg_n): average number of times each base being sequenced.</p>

<p>column 7 (cvg_wga): coverage of WGA sample. </p>
      </section>
    </div>
    <footer>
      <p>Project maintained by <a href="https://github.com/zhixingfeng">zhixingfeng</a></p>
      <p>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></p>
    </footer>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
    
  </body>
</html>