seqBMA
======
seqBMA is a toolkit for quantitative DNA modification detection from SMRT sequencing data.

System
======
Linux


Dependencies
============
1. R from http://www.r-project.org/
(Add directory containing binary file 'Rscript' to your enviroment variable PATH)

2. HDF5 from http://www.hdfgroup.org/HDF5/release/obtainsrc.html. 
(Specify enviroment varible HDF5_HOME as the directory of installed HDF5)

Installation
============
Unzip the downloaded file, enter the directory and type ./install.sh 


License
=======
GPL2


Input
=====
The Input of seqBMA is .cmp.h5 file(aligned reads) generated by SMRTanalysis developed by Pacific Biosciences.


Example
=======
http://bioinfo.au.tsinghua.edu.cn/member/zfeng/example_release.tar.gz

Maintainer
==========
Zhixing Feng


Contact
=======
zxfeng.thu@gmail.com


Usage
=====
There are 3 steps: pileup reads, learn prior and detect modification proportion.

Step 1: pileup reads. Firstly, we need to pileup reads position by position.

	seqBMA_pileup  [-r <reagent>] [-m <mapQVthreshold>] <cmpH5file> <outdir>

	Where:
   		-r <reagent>,  --reagent <reagent>
     		chemistry used for SMRT sequencing, candidates are: "C2", default is "C2"

   		-m <mapQVthreshold>,  --mapQVthreshold <mapQVthreshold>
     		minimal mapQV, default is 255

   		<cmpH5file>
     		(required)  aligned data in cmpH5 format

   		<outdir>
     		(required)  output directory

Step 2: learn prior. To reduce uncertainty of estimation, we need to roughly estimate IPD distribution of modified bases.

	seqBMA_prior  <native pileup dir> <WGA pileup dir> <outdir>

	Where:

   		<native pileup dir>
     		(required)  native pileup data directory

   		<WGA pileup dir>
     		(required)  WGA pileup data directory

   		<outdir>
     		(required)  output directory

Step 3: detect modification proportion. Now we fit a Bayesian mixture model using prior estimated in step 2.

	seqBMA_detect <native pileup dir> <WGA pileup dir> <prior dir> <outdir>

	Where:
   		<native pileup dir>
     		(required)  native pileup data directory

   		<WGA pileup dir>
     		(required)  WGA pileup data directory

   		<prior dir>
     		(required)  directory containing estimated prior

   		<outdir>
     		(required)  output directory

 





